File: app/src/main/java/org/fossasia/susi/ai/activities/SignUp/SignUpActivity.java
package org.fossasia.susi.ai.activities.SignUp;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.RadioButton;

import org.fossasia.susi.ai.R;
import org.fossasia.susi.ai.activities.ForgotPasswordActivity;
import org.fossasia.susi.ai.activities.LoginActivity;
import org.fossasia.susi.ai.helper.Constant;
import org.fossasia.susi.ai.helper.CredentialHelper;
import org.fossasia.susi.ai.helper.AlertboxHelper;


import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class SignUpActivity extends AppCompatActivity implements SignUpView{

    @BindView(R.id.email)
    protected TextInputLayout email;
    @BindView(R.id.password)
    protected TextInputLayout password;
    @BindView(R.id.confirm_password)
    protected TextInputLayout confirmPassword;
    @BindView(R.id.sign_up)
    protected Button signUp;
    @BindView(R.id.susi_default)
    protected RadioButton susiServer;
    @BindView(R.id.personal_server)
    protected RadioButton personalServer;
    @BindView(R.id.input_url)
    protected TextInputLayout url;

    private String alertTitle,alertMessage;
    SignUpInteractor signUpInteractor;


    private static final String TAG = "SignUpActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_up);
        ButterKnife.bind(this);
        signUpInteractor = new SignUpPresenterImpl(this, getApplicationContext());


        if(savedInstanceState!=null){
            email.getEditText().setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[0].toString());
            password.getEditText().setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[1].toString());
            confirmPassword.getEditText().setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[2].toString());
            if(savedInstanceState.getBoolean(Constant.SERVER)) {
                url.setVisibility(View.VISIBLE);
            } else {
                url.setVisibility(View.GONE);
            }
        }

        setupPasswordWatcher();
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        if(personalServer.isChecked()) {
            url.setVisibility(View.VISIBLE);
        } else {
            url.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        CharSequence values[] = {email.getEditText().getText().toString(), password.getEditText().getText().toString(), confirmPassword.getEditText().getText().toString() };
        outState.putCharSequenceArray(Constant.SAVED_STATES, values);
        outState.putBoolean(Constant.SERVER,personalServer.isChecked());
    }

    @Override
    public void onBackPressed(){
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(SignUpActivity.this);
        alertDialog.setCancelable(false);
        alertDialog.setMessage(R.string.error_cancelling_signUp_process_text);
        alertDialog.setPositiveButton(getApplicationContext().getString(R.string.yes), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                SignUpActivity.super.onBackPressed();
            }
        });
        alertDialog.setNegativeButton(getApplicationContext().getString(R.string.no),null);
        AlertDialog alert = alertDialog.create();
        alert.show();
        Button yes = alert.getButton(DialogInterface.BUTTON_POSITIVE);
        Button no = alert.getButton(DialogInterface.BUTTON_NEGATIVE);
        yes.setTextColor(getResources().getColor(R.color.md_blue_500));
        no.setTextColor(getResources().getColor(R.color.md_red_500));
    }

    @Override
    public void alertSuccess() {
        DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                finish();
            }
        };
        alertTitle = getResources().getString(R.string.signup);
        alertMessage = getResources().getString(R.string.signup_msg);
        AlertboxHelper successAlertboxHelper = new AlertboxHelper(SignUpActivity.this, alertTitle, alertMessage, dialogClickListener, null, getResources().getString(R.string.ok), null, getResources().getColor(R.color.md_blue_500));
        successAlertboxHelper.showAlertBox();
    }

    @Override
    public void alertError(String title,String message) {
        AlertboxHelper errorAlertboxHelper = new AlertboxHelper(SignUpActivity.this, title, message, null, null, getResources().getString(R.string.ok), null, Color.BLUE);
        errorAlertboxHelper.showAlertBox();
    }

    @Override
    public void setErrorEmail(String msg) {
        email.setError(msg);
    }

    @Override
    public void setErrorPass(String msg) {
        password.setError(msg);
    }

    @Override
    public void setErrorConpass(String msg) {
        confirmPassword.setError(msg);
    }

    @Override
    public void setErrorUrl(String msg) {
        url.setError(msg);
    }

    @Override
    public void enableSignUp(boolean bool) {
        signUp.setEnabled(bool);
    }

    @Override
    public boolean isPersonalServer() {
        return personalServer.isChecked();
    }

    @Override
    public void clearFiled() {
        CredentialHelper.clearFields(email, password, confirmPassword);
    }

    @Override
    public boolean checkIfEmptyUrl() {
        return CredentialHelper.checkIfEmpty(url,this);
    }

    @Override
    public boolean isURLValid() {
        return CredentialHelper.isURLValid(url,this);
    }

    @Override
    public boolean isEmailValid(String email) {
        return CredentialHelper.isEmailValid(email);
    }

    @Override
    public void setupPasswordWatcher() {
        password.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (!hasFocus) {
                    CredentialHelper.checkPasswordValid(password, SignUpActivity.this);
                }
            }
        });
    }

    @Override
    public String getValidURL() {
        return CredentialHelper.getValidURL(url, this);
    }

    @Override
    public ProgressDialog showProcess() {
        final ProgressDialog progressDialog = new ProgressDialog(SignUpActivity.this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage(getApplicationContext().getString(R.string.signing_up));
        progressDialog.show();
        return progressDialog;
    }

    @Override
    public boolean checkPasswordValid() {
        return CredentialHelper.checkPasswordValid(password, this);
    }

    @Override
    public boolean checkCredentials() {

        return CredentialHelper.checkIfEmpty(email, this) |
                CredentialHelper.checkIfEmpty(password, this) |
                CredentialHelper.checkIfEmpty(confirmPassword, this);
    }

    @Override
    public void alertFailure() {
        DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                Intent intent = new Intent(SignUpActivity.this, LoginActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
                finish();
            }
        };
        DialogInterface.OnClickListener dialogClickListenern = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                Intent intent = new Intent(SignUpActivity.this, ForgotPasswordActivity.class);
                startActivity(intent);
                finish();
            }
        };
        alertTitle = getResources().getString(R.string.error_email);
        alertMessage = getResources().getString(R.string.error_msg);
        AlertboxHelper failureAlertboxHelper = new AlertboxHelper(SignUpActivity.this, alertTitle, alertMessage, dialogClickListener, dialogClickListenern, getResources().getString(R.string.ok), getResources().getString(R.string.forgot_pass_activity), getResources().getColor(R.color.md_blue_500));
        failureAlertboxHelper.showAlertBox();
    }

    @OnClick(R.id.sign_up)
    public void signUp() {
            signUpInteractor.signUp(email.getEditText().getText().toString() , password.getEditText().getText().toString(), confirmPassword.getEditText().getText().toString());
        }

    @OnClick(R.id.personal_server)
    public void showURL() {
        url.setVisibility(View.VISIBLE);
    }

    @OnClick(R.id.susi_default)
    public void hideURL() {
        url.setVisibility(View.GONE);
    }

}

File: app/src/main/java/org/fossasia/susi/ai/signup/SignUpActivity.kt
package org.fossasia.susi.ai.signup

import android.app.AlertDialog
import android.app.Dialog
import android.app.ProgressDialog
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_sign_up.*
import org.fossasia.susi.ai.R
import org.fossasia.susi.ai.chat.ChatActivity
import org.fossasia.susi.ai.helper.AlertboxHelper
import org.fossasia.susi.ai.helper.Constant
import org.fossasia.susi.ai.helper.CredentialHelper
import org.fossasia.susi.ai.login.ForgotPass
import org.fossasia.susi.ai.login.LoginActivity
import org.fossasia.susi.ai.signup.contract.ISignUpPresenter
import org.fossasia.susi.ai.signup.contract.ISignUpView
import org.fossasia.susi.ai.skills.SkillsActivity
import org.koin.android.ext.android.inject
import org.koin.core.parameter.parametersOf

/**
 * <h1>The SignUp activity.</h1>
 * <h2>This activity is used to signUp into the app.</h2>
 *
 * Created by mayanktripathi on 05/07/17.
 */

class SignUpActivity : AppCompatActivity(), ISignUpView {

    private val signUpPresenter: ISignUpPresenter by inject { parametersOf(this) }
    private lateinit var progressDialog: ProgressDialog
    private lateinit var forgotPasswordProgressDialog: Dialog
    private lateinit var builder: AlertDialog.Builder
    private var checkDialog: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_up)
        setupPasswordWatcher()

        if (savedInstanceState != null) {
            email.editText?.setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[0].toString())
            password.editText?.setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[1].toString())
            confirmPassword.editText?.setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[2].toString())

            if (savedInstanceState.getBoolean(Constant.SERVER)) {
                inputUrlSignUp.visibility = View.VISIBLE
            } else {
                inputUrlSignUp.visibility = View.GONE
            }

            if (savedInstanceState.getBoolean(Constant.SAVE_DIALOG_STATE)) {
                onBackPressed()
            }
        }

        val bundle = intent.extras
        val string = bundle?.getString("email")
        if (string != null)
            email.editText?.setText(string)

        progressDialog = ProgressDialog(this@SignUpActivity)
        progressDialog.setCancelable(false)
        progressDialog.setMessage(this.getString(R.string.signing_up))

        builder = AlertDialog.Builder(this)
        builder.setView(R.layout.progress)
        forgotPasswordProgressDialog = builder.create()

        addListeners()

        cancelRequestPassword()
    }

    private fun addListeners() {
        showURL()
        signUp()
        signUpToLoginPage()
        cancelSignUp()
        signUpToTermsConditionPage()
        skipSignUp()
    }

    fun skipSignUp() {
        skipSignUp.setOnClickListener {
            val dialogClickListener = AlertDialog.Builder(this)
            with(dialogClickListener) {
                setTitle(R.string.dialog_skip_sign_up)
                setMessage(R.string.error_skipping_signUp_process_text)
                setPositiveButton(R.string.yes, DialogInterface.OnClickListener { dialog, id ->
                    val intent = Intent(this@SignUpActivity, ChatActivity::class.java)
                    startActivity(intent)
                    finish()
                })
                setNegativeButton(android.R.string.no, DialogInterface.OnClickListener {
                    dialog, id -> dialog.cancel()
                })
                show()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val values = arrayOf<CharSequence>(email.editText?.text.toString(), password.editText?.text.toString(), confirmPassword.editText?.text.toString())
        outState.putCharSequenceArray(Constant.SAVED_STATES, values)
        outState.putBoolean(Constant.SERVER, customServerSignUp.isChecked)
        outState.putBoolean(Constant.SAVE_DIALOG_STATE, checkDialog)
    }

    override fun onBackPressed() {
        val dialogClickListener = DialogInterface.OnClickListener { _, _ -> super@SignUpActivity.onBackPressed() }
        val alertMessage = getString(R.string.error_cancelling_signUp_process_text)
        val dialogTitle = getString(R.string.dialog_cancel_sign_up)
        val successAlertboxHelper = AlertboxHelper(this@SignUpActivity, dialogTitle, alertMessage, dialogClickListener, null,
                resources.getString(R.string.cancel), resources.getString(R.string.stay_here), resources.getColor(R.color
                .md_blue_500))
        successAlertboxHelper.showAlertBox()
        checkDialog = true
    }

    override fun alertSuccess() {
        val dialogClickListener = DialogInterface.OnClickListener { _, _ -> finish() }
        val alertTitle = getString(R.string.signup)
        val alertMessage = getString(R.string.signup_msg)
        val successAlertboxHelper = AlertboxHelper(this@SignUpActivity, alertTitle, alertMessage, dialogClickListener, null, resources.getString(R.string.ok), null, resources.getColor(R.color.md_blue_500))
        successAlertboxHelper.showAlertBox()
    }

    override fun alertFailure() {
        val dialogClickListener = DialogInterface.OnClickListener { _, _ ->
            val intent = Intent(this@SignUpActivity, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(intent)
            finish()
        }
        val dialogClickListener1 = DialogInterface.OnClickListener { _, _ ->
            val email1 = email.editText?.text.toString()
            val isPersonalServerChecked = customServerSignUp.isChecked
            val url = inputUrlSignUp.editText?.text.toString()
            email.error = null
            inputUrlSignUp.error = null
            signUpPresenter.requestPassword(email1, url, isPersonalServerChecked)
        }
        val alertTitle = getString(R.string.error_email)
        val alertMessage = getString(R.string.error_msg)
        val failureAlertboxHelper = AlertboxHelper(this@SignUpActivity, alertTitle, alertMessage, dialogClickListener, dialogClickListener1, resources.getString(R.string.ok), resources.getString(R.string.forgot_pass_activity), resources.getColor(R.color.md_blue_500))
        failureAlertboxHelper.showAlertBox()
    }

    override fun setErrorConpass(msg: String) {
        confirmPassword?.error = msg
    }

    override fun enableSignUp(bool: Boolean) {
        signUp?.isEnabled = bool
    }

    override fun clearField() {
        CredentialHelper.clearFields(email, password, confirmPassword)
    }

    override fun showProgress(bool: Boolean) {
        if (bool) progressDialog.show() else progressDialog.dismiss()
    }

    override fun invalidCredentials(isEmpty: Boolean, what: String) {
        if (isEmpty) {
            when (what) {
                Constant.EMAIL -> email.error = getString(R.string.email_cannot_be_empty)
                Constant.PASSWORD -> password.error = getString(R.string.password_cannot_be_empty)
                Constant.INPUT_URL -> inputUrlSignUp.error = getString(R.string.url_cannot_be_empty)
                Constant.CONFIRM_PASSWORD -> confirmPassword.error = getString(R.string.field_cannot_be_empty)
                Constant.ACCEPT_TERMS_AND_CONDITIONS -> {
                    acceptTermsAndConditions.error = getString(R.string.accept_terms_and_conditions)
                    Toast.makeText(this@SignUpActivity, R.string.accept_terms_and_conditions, Toast.LENGTH_SHORT).show()
                }
            }
        } else {
            when (what) {
                Constant.EMAIL -> email.error = getString(R.string.invalid_email)
                Constant.INPUT_URL -> inputUrlSignUp.error = getString(R.string.invalid_url)
                Constant.PASSWORD -> password.error = getString(R.string.error_password_matching)
            }
        }
    }

    override fun passwordInvalid() {
        password.error = getString(R.string.pass_validation_text)
        signUp.isEnabled = true
    }

    private fun showURL() {
        customServerSignUp.setOnClickListener { inputUrlSignUp.visibility = if (customServerSignUp.isChecked) View.VISIBLE else View.GONE }
    }

    private fun setupPasswordWatcher() {
        password.editText?.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            password.error = null
            if (!hasFocus)
                signUpPresenter.checkForPassword(password.editText?.text.toString())
        }
    }

    private fun signUpToLoginPage() {
        signUpToLogin.setOnClickListener {
            signUpPresenter.loginLogout()
            val intent = Intent(this@SignUpActivity, LoginActivity::class.java)
            intent.putExtra("email", email.editText?.text.toString())
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(intent)
        }
    }

    private fun signUpToTermsConditionPage() {
        signUpToTermsCondition.setOnClickListener {
            val intent = Intent(this@SignUpActivity, SkillsActivity::class.java)
            intent.putExtra(Constant.SIGN_UP_TO_PRIVACY, true)
            startActivity(intent)
        }
    }

    private fun cancelSignUp() {
        progressDialog.setOnCancelListener {
            signUpPresenter.cancelSignUp()
            signUp.isEnabled = true
        }
    }

    override fun onSignUpError(title: String?, message: String?) {
        val notSuccessAlertboxHelper = AlertboxHelper(this@SignUpActivity, title, message, null, null, getString(R.string.ok), null, Color.BLUE)
        notSuccessAlertboxHelper.showAlertBox()
        signUp.isEnabled = true
    }

    private fun signUp() {

        signUp.setOnClickListener {

            email.error = null
            password.error = null
            confirmPassword.error = null
            inputUrlSignUp.error = null

            val stringEmail = email.editText?.text.toString()
            val stringPassword = password.editText?.text.toString()
            val stringConfirmPassword = confirmPassword.editText?.text.toString()
            val stringURL = inputUrlSignUp.editText?.text.toString()

            signUpPresenter.signUp(stringEmail, stringPassword, stringConfirmPassword, !customServerSignUp.isChecked, stringURL, acceptTermsAndConditions.isChecked)
        }
    }

    override fun showForgotPasswordProgress(boolean: Boolean) {
        if (boolean) forgotPasswordProgressDialog.show() else forgotPasswordProgressDialog.dismiss()
    }

    override fun resetPasswordSuccess() {
        startActivity(Intent(this@SignUpActivity, ForgotPass::class.java))
    }

    override fun resetPasswordFailure(title: String?, message: String?, button: String?, color: Int) {
        val notSuccessAlertboxHelper = AlertboxHelper(this@SignUpActivity, title, message, null, null, button, null, color)
        notSuccessAlertboxHelper.showAlertBox()
    }

    private fun cancelRequestPassword() {
        progressDialog.setOnCancelListener {
            signUpPresenter.cancelSignup()
        }
    }
}
